util/pathstrings_test.go

```diff
@@ -0,0 +1,290 @@
+// Copyright 2020 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package util
+
+import (
+	"strings"
+	"testing"
+
+	"google.golang.org/protobuf/encoding/prototext"
+	"google.golang.org/protobuf/proto"
+
+	gnmipb "github.com/openconfig/gnmi/proto/gnmi"
+)
+
+// XXX: This is copied code from ygot package. ygot's code should probably
```

Comment:
Is there a reason to take on this debt in the codebase rather than fix it now? It seems like there'd be relatively minimal refactoring here -- i.e., we move the actual implementation to util, and keep the public APIs in ygot just calling these util functions.

Reply:
We maybe also should consider whether `util` itself is going to grow into a similar situation, and create a smaller package just for path handling functions.

Reply:
From a quick scan of the internal dependencies it looks like it's possible to split out all **gnmi**-path related functions into a new package. This is all of `gnmi.go` and the `GetNode` and relatives in `reflect.go`.

If we move **all** path-related functions into a new package, we'll essentially be moving `util` to a new package.

So, I'm not sure if we can quite create a package called `pathutil`, but `gnmiutil` is still possible.

Reply:
After looking at the internal dependencies further, I'm still not seeing why we should split the path functions into its own package.

The `util` package has roughly the following high-level components in topological/dependency order:
1. Debug utilities
2. Basic error type
3. ygot-specific goyang schema helpers
4. string path helpers operating on the goyang schema, which uses 3
5. gNMI path helpers, which uses 4
 - 4 can be separated into helpers that don't use the goyang schema (4a) vs. do (4b), and 5 only depends on those that don't (4b); however, 4a&4b are so intertwined that 4b also depends on 4a's internal methods, so it doesn't make sense functionally to split 4 into two.
6. GoStruct helpers
7. GoStruct helpers that operating on goyang's schema, which uses 4
8. GoStruct traversal functions based on goyang's schema, which uses 4, 6-7
9. GoStruct traversal functions based on gNMI paths and goyang's schema, that uses 3-7.

While 5&9 can be separated into another package called `gnmiutil`, which would contain pathstrings.go as well, it would be splitting based on that they both use `gpb.Path` -- the meaning of `gnmiutil` would still be unclear to the reader, which I think is the true problem: that when a reader reads `util`, they don't know what it does.

I think it makes sense, rather, to split `util` into the following packages in topological order:
- `util`: 1-3
- `pathutil`: 4-5 + ygot/pathstrings.go
- `gostructutil` 6-9
What do you think @robshakir

Reply:
I think this set sounds a good set to have. Apologies for the delay in responding here. I'm happy with this split, which AIUI should mean that we don't need to duplicate code here.

Reply:
Since generic package names are not recommended, perhaps a better set of names would be
`internal`: 1 https://golang.org/doc/go1.4#internalpackages
`ygoterrors`: 2
`yangschema`: 3 (we have the option of simplifying some of the function names, e.g. `util.SchemaPath` -> `yangschema.Path`)
`yangpath`: 4-5 + ygot/pathstrings.go
`gostruct`: 6-9

We might also consider renaming `genutil` to `ygotgen`.

Reply:
Part of util refactoring effort: https://github.com/openconfig/ygot/issues/556

Reply:
Ah, sorry -- I thought that this was what you had decided upon. I think that this sounds a reasonable split -- I'm all OK with dividing out to very specific packages, I think in general it means that we end up with a cleaner import set from other modules.